feat(cart,post): 장바구니, 의뢰글 모듈 기본 설정

sora0319 opened the pull request:
## 📌 목적
- 장바구니, 의뢰글 통합 모듈에서 api 개발을 위한 기본적인 틀과 테스트 코드 틀 생성
- 안정적인 DB연결과 테이블 생성을 위한 schema.sql을 생성

## ✅ 변경 사항
- swagger 적용
- cart controller, service, repository 생성
- post(commissions) controller, service, repository 생성
- CartService, CommissionsService 단위 테스트 코드 기본 틀 작성
- schema.sql 로 자동적으로 테이블 생성 및 JPA 가 테이블 생성에 관여하지 않도록 설정

## 🧪 테스트 방법
- API는 기본 틀만 있어서 테스트 시 test ok 로 나옵니다

## 📎 참고 사항
- resourses/secham.sql을 설정을 하여서 JPA가 테이블 생성에 관여하지 않고 자동적으로 테이블이 생성되도록 설정하였습니다 cartpost-servie의 yml을 참고하여서 적용하시면 됩니다
- swagger는 http:localhost:8000/api/commissions/swagger-ui 로 접속하면 아래와 같이 api가 보이게 됩니다
<img width="1498" height="707" alt="image" src="https://github.com/user-attachments/assets/466f9b30-5703-41f5-b3f8-975c0297afeb" />


## 📝 제출자 검토 리스트

- [x] 실행 시 오류가 없나요?
- [x] 테스트 전부 통과 했나요?
- [x] 코드 컨벤션을 모두 지켰나요?
- [x] 브랜치가 develop에 merge 요청을 보냈나요?

## 📝 검토자 체크 리스트

> [!IMPORTANT]
> 직접 모든 부분에 대해 검토하고 멘션이나 이모티콘 남겨주세요

- [ ] 요청을 받은 후 하루 이내에 피드백을 보내주세요
- [ ] 본인이 검토해야하는 PR 피드백을 주세요.
- [ ] 브랜치가 develop에 merge 요청을 보냈나요?

sora0319 committed:
feat(cart): 장바구니 모듈 의존성 추가

Lombok, JPA, H2, Swagger 의존성 추가함

sora0319 committed:
feat(cart,post): swagger API 문서 기본 설정

sora0319 committed:
feat(cart,post): 자동 날짜 업데이트 JPA 기능 추가

sora0319 committed:
feat(cart,post): BaseEntity 추가

sora0319 committed:
feat(cart,post): ResponseDto 추가

sora0319 committed:
feat(cart,post): 빈 데이터 값을 위한 EmptyDto

sora0319 committed:
feat(cart): cart controller를 위한 swagger 파일 추상화 생성

sora0319 committed:
chore(cart,post): code convention에 맞추어 수정

sora0319 committed:
feat(cart,post): 자동으로 dto 조회하도록 defaultValue 삭제

sora0319 committed:
feat(cart,post): 예외 처리를 위한 공통 예외 생성

sora0319 committed:
feat(cart,post): 예외 처리를 위한 예외 처리 코드 모음 생성

sora0319 committed:
feat(cart,post): valid 유효성 검사 에러를 관리하기 위한 정보 객체 생성

sora0319 committed:
feat(cart,post): 응답 dto에 에러 처리를 위한 팩토리 메소드 생성

sora0319 committed:
feat(cart,post): 전역 에러 관리 handler 생성

sora0319 committed:
chore(cart,post): 전역 에러 관리 handler return value 이름 수정

sora0319 committed:
chore(cart,post): EmptyDto를 EmptyResponse로 이름 변경

sora0319 committed:
feat(cart,post): swagger schema 추가

sora0319 committed:
chore(cart,post): EmptyDto를 EmptyResponse로 변경

sora0319 committed:
feat(cart): 아이템 목록 반환 response 생성

sora0319 committed:
feat(cart): 장바구니 controller 생성

sora0319 committed:
feat(cart, post): lombok 적용 코드 추가

sora0319 committed:
chore(cart, post): 사용 안하는 import 제거

sora0319 committed:
feat(cart, post): 컬럼 속성 명시 추가

sora0319 committed:
feat(cart): cart entity 추가

sora0319 committed:
feat(cart): cart item entity 추가

sora0319 committed:
feat(cart): cart repository 추가

sora0319 committed:
feat(cart): cart service 추가

sora0319 committed:
feat(cart): cart service 파라미터 변경

sora0319 committed:
feat(cart): cart service response dto 생성

sora0319 committed:
feat(cart): cart service response dto 생성

sora0319 committed:
feat(cart): 타입 record 로 변경

sora0319 committed:
feat(cart): item code 제거

sora0319 committed:
feat(cart): cart code 추가

sora0319 committed:
feat(post): 지불 방식 타입 추가

sora0319 committed:
feat(post): 의뢰글 태그 엔티티 생성

sora0319 committed:
feat(post): 의뢰글 엔티티 생성

sora0319 committed:
feat(post): commissions repository 생성

sora0319 committed:
feat(post): commissions tag repository 생성

sora0319 committed:
chore(cart,post): 날짜 변수명 변경

sora0319 committed:
chore(post): 의뢰글 controller dto 생성

sora0319 committed:
feat(post): 의뢰글 controller dto 생성

sora0319 committed:
chore(post): 의뢰글 service dto 생성

sora0319 committed:
chore(post): 의뢰글 service dto 생성

sora0319 committed:
chore(post): 의뢰글 service dto 생성

sora0319 committed:
chore(post): 의뢰글 service dto 생성

sora0319 committed:
chore(post): 의뢰글 controller 메서드 생성 및 swagger 문서용 분리

sora0319 committed:
chore(post): 의뢰글 controller 메서드 생성 및 swagger 문서용 분리

sora0319 committed:
feat(post): 의뢰글 service 생성

sora0319 committed:
chore(post): 의뢰글 service 생성

sora0319 committed:
feat(post): 의뢰글 service 생성

sora0319 committed:
Merge remote-tracking branch 'origin/feat/cart/7' into feat/cart/7

sora0319 committed:
Merge remote-tracking branch 'origin/feat/cart/7' into feat/cart/7

sora0319 committed:
chore(cart): 파일 이름 복수형 수정

sora0319 committed:
feat(cart): cart service 구현체 생성

sora0319 committed:
chore(cart): cart service 구현체 위치 변경

sora0319 committed:
test(cart): cart service 단위 테스트 기본 틀 구현

sora0319 committed:
test(post): 의뢰글 service 테스트 기본 틀 생성

sora0319 committed:
feat(post): 의뢰글 service 구현체 기본 틀 생성

sora0319 committed:
feat(common): 전체 DDL sql 문 생성

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/controller/CartController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Override
+    @DeleteMapping("/{item_code}")
```
아이템 리소스에 대한 삭제이니 `/items/{item_code}` 는 어떨지 건의 드려봅니다.

sora0319 replied:
그게 더 명확한 방법일 것 같네요! 감사합니다!

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/model/CartItemsEntity.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Table(name = "cart_itmes")
```
오타가 있습니다!

sora0319 replied:
헉, 수정하겠습니다

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/model/CartsEntity.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Table(name = "cart_itmes")
```
`carts`여야 할 것 같습니다.

sora0319 replied:
현재 items에 복수형이라서 cart_items로 했고, 1개의 장바구니에 대한 아이템들이니 이게 더 명확할 것 같습니다

추가적으로 현재 cart, cart item이 같은 모듈 안에서 1: N 관계로 되어 있는데 이 두 테이블이 서로 다른 모듈로 나누어 질 일이 없습니다. 그러나 현재 반정규화가 되어있어서 애플리케이션 로직에서 이 무결성을 모두 책임져야 하므로, 로직이 복잡해지고 버그가 발생하기 쉬운 상태라 같은 모듈에 있는 관계가 있는 테이블에 대해서는 foreign key 설정을 하는 것이 어떤가요?

ZZAMBAs replied:
여기는 장바구니에 대한 테이블인 것 같아서 `carts`로 제안드렸습니다. 
FK 설정은 물리적으로 걸기도 하고 편의를 위해 안 걸기도 한다는 것 같습니다. 이전에 회원과 글을 FK로 설정하면 모든 글이 지워지기 전에 회원을 삭제할 수 없는 불편함이 있다고 들은 적 있고, 기록용으로 남겨둔다는 말도 들었던 것 같아요.

sora0319 replied:
cart item eneity 인줄 착각했네요
이 주제에 대해서는 최종 들어가기 전에 한번 의논해 보는 것이 좋을 것 같습니다!!

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/commissions/controller/CommissionsApi.java:
```diff
@@ -0,0 +1,103 @@
+            @Parameter(name = "X-CODE", description = "사용자 식별 코드", required = true, in = ParameterIn.HEADER),
+            @Parameter(name = "page", description = "페이지 번호 (0부터 시작)", example = "0"),
+            @Parameter(name = "size", description = "페이지 크기", example = "10"),
+            @Parameter(name = "sort", description = "정렬 기준 (예: createdAt,desc)", example = "isOpen,asc")
```
제 개인적인 의견이지만, 각각 받을 수 있는 요소는 인자를 따로 받는 게 괜찮다고 생각합니다.

sora0319 replied:
정렬 기준은 다수로 될 수 있다고 생각하여서 이렇게 하였습니다
만약 따로 받는다고 하면
createsort=createdAt&sort=desc&updatesort=updatedAt
이렇게 설정하는 것을 생각해 볼 수 있는데 가독성도 떨어지고, sort 조건이 늘어날 수록 관리하기 어려워진다고 생각하였습니다

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/commissions/model/vo/PaymentType.java:
```diff
@@ -0,0 +1,13 @@
+@Getter
+@RequiredArgsConstructor
+public enum PaymentType {
+    MONTH("월급"), PER_JOB("건당");
```
저 같은 경우는 ERD를 보고 각각 `MONTHLY`, `ONE_TIME` 으로 작성했습니다. 통일하는 것은 어떻게 생각하시나요?

sora0319 replied:
그러면 MONTHLY, PER_JOB 으로 하는 것은 어떤 가요?
이렇게 하는게 더 의미가 명확할 것 같아서 의견 물어봅니다!

ZZAMBAs replied:
네, 괜찮은 것 같습니다. 현재 하은님 코드에서도 건당이 `ONE_CASE`로 되어 있음을 확인해서 슬랙에 따로 남기겠습니다.

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/commissions/repository/CommissionsRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CommissionsRepository extends JpaRepository<CommissionsEntity, String> {
```
Id 값은 code가 아닌, Long id를 사용하기 때문에 `Long`으로 변경해야 할 것 같습니다.

sora0319 replied:
헉!! 감사합니다
오류가 발생할 뻔 했내요

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/common/dto/ResponseDto.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    //@Valid 유효성 검사 에러 응답 (MethodArgumentNotValidException)
+    public static ResponseDto<List<FieldErrorDetail>> of(ErrorCode errorCode, BindingResult bindingResult) {
```
메서드 이름 `of`로는 잘 의미가 안보일 수도 있을 것 같아요. `createInvalidResponse` 같은 메서드 명은 어떻게 생각하세요?

sora0319 replied:
of 라는 메서드 이름이 '객체를 만든다' 라는 의미가 있습니다.
이러한 of 이름 규칙이 현재 자바 내의 메서드에서도 사용되고 있어서 통일성 있게 사용하기 위해 선택을 하였습니다
즉 ResponseDto.of() 는 괄호 안에 있는 값들도 ResponseDto 객체를 생성한다 라는 의미를 주기 위해 사용하였습니다

ZZAMBAs replied:
넵. 말씀하신대로 `of()`는 보편적인 의미입니다. 그런데 이 `ResponseDto`가 제네릭 타입임을 저는 염두해서 말씀드린 것이긴 합니다. 같은 `of()` 메서드지만 `ResponseDto<EmptyResponse>` 반환 타입을 갖는 `of()`와 이 `of()`는 다른 의미를 지닌다고 생각했어요. 그래서 구체적 이름을 사용하면 어떨까 싶어 제안 드렸습니다!

sora0319 replied:
반환 타입까지는 생각하지 못했네요!!
현재는 오류에 대한 변환 메서드들만 있지만, 나중에 성공에 대한 변환 메서드들이 생기면 이 또한 메서드 네이밍에 대해 역할을 잘 나타내지 못할 것 같네요
메서드 명을 변경하겠습니다!

ZZAMBAs commented on cartpost-service/src/main/java/com/example/cartpostservice/common/exception/CustomStatusCode.java:
```diff
@@ -0,0 +1,16 @@
+@Getter
+@RequiredArgsConstructor
+public enum ErrorCode {
+    SUCCESS(HttpStatus.OK, 0, "성공");
```
에러 코드에 성공이 있는 모양이 조금 어색하게 느껴져서 저는 `ResponseDto`에 따로 `ok()` 라는 정적 메서드를 만들기도 했는데, 어떻게 생각하시는지 궁금합니다!

sora0319 replied:
HttpStatus에서도 ok, badrequest, internal server error 등 어려 오류와 성공에 대한 값들이 같이 관리되고 있어서 이렇게 구성하는 것을 생각하였습니다
또한 만약 성공 코드를 따로 분리한다면, 성공에 대해서도 200,201,204 등 다양한 상태에 대해 관리하는 enum이 또 필요하다고 생각이 들어서 같이 관리하는 것이 더 효율적일 것 같습니다.

추가적으로 ErrorCode로 클래스를 명명하기 보다는 CustomStatusCode로 클래스 이름을 바꾸는 것이 현재 기능을 유지하면서 의미 전달에 더 명확할 것 같습니다.

ZZAMBAs replied:
현재는 에러 코드만 저희가 관리하고 있기 때문에 성공 코드 분리를 안해도 될 것 같고, 말씀하신대로 `CustomStatusCode`로 바꾼다면 괜찮을 것 같습니다!

ZZAMBAs reviewed (commented):
메모리 문제와 배송 이슈 등 불운이 좀 있으셨던 것 같습니다.. 코드 작성 고생하셨습니다!

j1suk1m commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/controller/CartController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Override
+    @DeleteMapping("/{item_code}")
```
컨벤션 정할 때 [케밥 케이스](https://developer.mozilla.org/en-US/docs/Glossary/Kebab_case)를 따르기로 했었습니다!
```suggestion
    @DeleteMapping("/{item-code}")
```

sora0319 replied:
오타가... 많네요..ㅠㅠ
수정하겠습니다

j1suk1m commented on cartpost-service/src/main/java/com/example/cartpostservice/commissions/controller/CommissionsController.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    @GetMapping("/total")
```
전체 커미션을 조회하는 API라면 `GET /api/commissions`만으로도 충분해 보입니다!
이미 단건 조회 API가 `GET /api/commissions/{commission-code}`로 구현되어 있어서, 이 방식이 전체적인 API 구조에도 더 자연스럽게 어울리지 않을까 생각이 들었습니다.

sora0319 replied:
uri 매핑 구조에도 그것에 더 직관적일 것 같네요
의견 감사합니다
수정하겠습니다

j1suk1m commented on cartpost-service/src/main/java/com/example/cartpostservice/commissions/controller/CommissionsController.java:
```diff
@@ -0,0 +1,68 @@
+            @RequestHeader("X-CODE") String code,
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "10") int size,
+            @RequestParam(defaultValue = "isOpen,asc") String sort) {
```
컨트롤러에서 `Pageable`을 파라미터로 받고, `@PageableDefault`와 함께 사용하는 방법도 있다고 알고 있습니다.

sora0319 replied:
아!! 그 방법을 잊어버렸었네요
그것이 코드 관리 측면에서도 더 유용할 것 같습니다!!

j1suk1m commented on cartpost-service/src/main/java/com/example/cartpostservice/commissions/service/CommissionsService.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.web.bind.annotation.RequestParam;
+
+@Service
+public interface CommissionsService {
```
소정님은 서비스 인터페이스도 따로 만드시는군요👀

sora0319 replied:
만약 현재 마감 기능이 api 요청을 받아서 처리하는 것으로 구현이 되어 있는데 추가기능으로 자동으로 마감되게 하게 하려고 한다고 가정을 합니다.
이때 기존 코드를 삭제하여서 개발하는 것보다 새로운 구현체로 자동 마감 기능을 구현하여서 선택 마감과, 자동 마감 기능 2가지를 모두 제시할 수 있고, 컨트롤러에 대한 의존성 (SOLID)가 만족할 수 있습니다.

이러한 여러 변화에 대해 대응하기 위해 미리 인터페이스로 구현을 하고, 최대한 의존성을 낮추려고 인터페이스를 사용하고 있습니다

j1suk1m reviewed (commented):
수고 많으셨습니다!

sora0319 committed:
feat(cart,post): 삭제 상태 초기화 및 상태 변경 메서드 추가

sora0319 committed:
refactor(cart,post): ErrorCode를 CustomStatusCode로 이름 변환

기존 ErrorCode로 명명되어 있을 때에는 성공 시 반환하는 코드에 대한 역할이 부자연스러웠습니다. 이러한 문제점을 해결하기 위해 이름을 CustomStatusCode로 변경하여 성공, 실패에 대한 상태 코드를 관리하도록 하였습니다

sora0319 committed:
refactor(cart): uri의 명확성을 위해 수정

sora0319 committed:
fix(cart): 테이블 이름 오류 수정

sora0319 committed:
refactor(post): 기존 파라미터에 페이지 관리 파라미터를 입력 받았던 것을 Pageable로 관리하도록 수정

기존에는 페이지 요청에 따른 파라미터들을 각각 입력 받았지만, Pageable로 변경하면서

sora0319 committed:
fix(post): ID가 Long 기준으로 되어 있는데 code를 기준으로 String으로 등록된 것을 수정

sora0319 committed:
refactor(post): MONTHLY, PER_JOB으로 각 금액 지불 방식 이름 명명 수정

programofktw commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/model/CartItemsEntity.java:
```diff
@@ -0,0 +1,32 @@
+@Table(name = "cart_items")
+@Entity
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
이 후 구현간 프로덕트 코드에서 사용될 생성자가 추가되어야할 것 같아요!

sora0319 replied:
넵 추가하였습니다!

programofktw commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/repository/CartItemsRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
해당 Repository에서도 Long 대신 String이 들어가 있습니다!

programofktw commented on cartpost-service/src/main/java/com/example/cartpostservice/cart/repository/CartsRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CartsRepository extends JpaRepository<CartsEntity, String> {
```
해당 클래스에서 ID 타입이 Long 대신 STring으로 작성되어 있습니다!
인수님이 먼저 말씀해주셨지만 혹시나 해서 Comment 남겨둘게요!

programofktw reviewed (commented):
Repository의 ID 설정이나 Entity 프로덕션 코드에서 사용할 생성자를 제외하고는 크게 문제 될 코드가 없어보입니다! 
이후 작업에서 수정하셔도 되서 approve 해둘게요!

programofktw reviewed (approved):
장비 이슈로 힘드셨을 텐데 고생하셨습니다!
